discoverd/server/handler.go

```diff
@@ -0,0 +1,287 @@
+
+	// Remove instance from the store.
+	if err := h.Store.RemoveInstance(service, instanceID); IsNotFound(err) {
+		hh.ObjectNotFoundError(w, err.Error())
+		return
+	} else if err != nil {
+		hh.Error(w, err)
+		return
+	}
+}
+
+// serveGetInstances returns a list of all instances for a service.
+func (h *Handler) serveGetInstances(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+	// If the client is requesting a stream, then handle as a stream.
+	if strings.Contains(r.Header.Get("Accept"), "text/event-stream") {
+		h.serveStream(w, params, discoverd.EventKindUp|discoverd.EventKindUpdate|discoverd.EventKindDown)
+		return
+	}
+
+	// Otherwise read instances from the store.
+	instances := h.Store.Instances(params.ByName("service"))
+	if instances == nil {
+		hh.ObjectNotFoundError(w, "service not found")
+		return
+	}
+
+	// Write instances to the response.
+	hh.JSON(w, 200, instances)
+}
+
+// servePutLeader sets the leader for a service.
+func (h *Handler) servePutLeader(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+	// Retrieve path parameters.
+	service := params.ByName("service")
+
+	// Check if the service allows manual leader election.
+	config := h.Store.Config(service)
+	if config == nil || config.LeaderType != discoverd.LeaderTypeManual {
+		hh.ValidationError(w, "", "service leader election type is not manual")
+		return
+	}
+
+	// Read instance from the request.
+	inst := &discoverd.Instance{}
+	if err := hh.DecodeJSON(r, inst); err != nil {
+		hh.Error(w, err)
+		return
+	}
+
+	// Manually set the leader on the service.
+	if err := h.Store.SetLeader(service, inst.ID); err != nil {
+		hh.Error(w, err)
+		return
+	}
+}
+
+// serveGetLeader returns the current leader for a service.
+func (h *Handler) serveGetLeader(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+	// Process as a stream if that's what the client wants.
+	if strings.Contains(r.Header.Get("Accept"), "text/event-stream") {
+		h.serveStream(w, params, discoverd.EventKindLeader)
+		return
+	}
+
+	// Otherwise retrieve the current leader.
+	service := params.ByName("service")
+	leader := h.Store.Leader(service)
+	if leader == nil {
+		hh.ObjectNotFoundError(w, "no leader found")
+		return
+	}
+
+	// Write leader to the response.
+	hh.JSON(w, 200, leader)
+}
+
+// servePing returns a 200 OK.
+func (h *Handler) servePing(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+}
+
+// serveStream creates a subscription and streams out events in SSE format.
+func (h *Handler) serveStream(w http.ResponseWriter, params httprouter.Params, kind discoverd.EventKind) {
+	// Create a buffered channel to receive events.
+	ch := make(chan *discoverd.Event, StreamBufferSize)
+
+	// Subscribe to events on the store.
+	service := params.ByName("service")
+	stream := h.Store.Subscribe(service, true, kind, ch)
+
+	// Create and serve an SSE stream.
+	s := sse.NewStream(w, ch, nil)
+	s.Serve()
+	s.Wait()
```

Comment:
I tried using `sse.ServeStream` but it doesn't give me the stream reference to use with `s.CloseWithError()` a couple lines below.